### ISSUE TITLE
invalid caveats

### ISSUE THREAD

**Issue by Geal** [CleverCloud/biscuit-pulsar#8]
in the verifier caveats:

facts like `topic_operation` should be ambient facts

the `canProduceAsync` caveat only checks that the `topic_operation` fact is there:
https://github.com/CleverCloud/biscuit-pulsar/blob/ad2b3db3f65b37cfc6fadca40550d61766545992/biscuit-pulsar/src/main/java/com/clevercloud/biscuitpulsar/AuthorizationProviderBiscuit.java#L173-L178

It should do something like this:

```java
// add the topic_operation ambient fact
verifier.add_fact(fact("topic_operation",
  Arrays.asList(s("ambient"), string(topicName.getTenant()), string(topicName.getNamespacePortion()), string(topicName.getLocalName()), s("produce"))));
// add the topic ambient fact
verifier.add_fact(fact("topic",
  Arrays.asList(s("ambient"), string(topicName.getTenant()), string(topicName.getNamespacePortion()), string(topicName.getLocalName()))));
// check that we have the right to produce on that topic
verifier.add_caveat(new Caveat(Arrays.asList(
  rule("check_right",
    Arrays.asList(),
    Arrays.asList(pred("right",
      Arrays.asList(s("authority"), s("topic"), string(topicName.getTenant()), string(topicName.getNamespacePortion()), string(topicName.getLocalName()), s("produce"))))
))));
```
`canConsumeAsync` has the same issue:
https://github.com/CleverCloud/biscuit-pulsar/blob/ad2b3db3f65b37cfc6fadca40550d61766545992/biscuit-pulsar/src/main/java/com/clevercloud/biscuitpulsar/AuthorizationProviderBiscuit.java#L207-L211

`allowNamespaceOperationAsync` as well:
https://github.com/CleverCloud/biscuit-pulsar/blob/ad2b3db3f65b37cfc6fadca40550d61766545992/biscuit-pulsar/src/main/java/com/clevercloud/biscuitpulsar/AuthorizationProviderBiscuit.java#L440-L444

same for `allowNamespacePolicyOperationAsync`:
https://github.com/CleverCloud/biscuit-pulsar/blob/ad2b3db3f65b37cfc6fadca40550d61766545992/biscuit-pulsar/src/main/java/com/clevercloud/biscuitpulsar/AuthorizationProviderBiscuit.java#L490-L494

also, why is the token deserialized again here:
https://github.com/CleverCloud/biscuit-pulsar/blob/ad2b3db3f65b37cfc6fadca40550d61766545992/biscuit-pulsar/src/main/java/com/clevercloud/biscuitpulsar/AuthorizationProviderBiscuit.java#L240-L249

please add more unit tests with tokens that do not have the required rights, they would have caught those issues

**Comment by KannarFr** [comment]
@Geal I tried to apply your first suggestion on `canProduce`, got the following error:

```
02:37:57.736 [main] DEBUG com.clevercloud.biscuitpulsar.AuthorizationProviderBiscuitTest - Biscuit {
	symbols: [authority, ambient, resource, operation, right, current_time, revocation_id, admin, namespace, ns_operation, backlog_write, deduplication_read, retention_read, clear_backlog, offload_write, max_consumers_write, replication_write, replication_rate_read, unsubscribe, encryption_write, rate_write, ttl_read, get_permission, backlog_read, delayed_delivery_read, subscription_auth_mode_read, all_write, create_topic, delete_topic, max_unacked_write, add_bundle, anty_affinity_write, persistence_write, replication_rate_write, max_consumers_read, max_producers_read, schema_compatibility_strategy_read, deduplication_write, persistence_read, schema_compatibility_strategy_write, ttl_write, retention_write, anty_affinity_read, all_read, max_producers_write, get_topic, replication_read, encryption_read, get_bundle, get_topics, grant_permission, compaction_write, rate_read, delete_bundle, revoke_permission, compaction_read, offload_read, max_unacked_read, delayed_delivery_write, subscription_auth_mode_write, topic, topic_operation, lookup, consume, produce, limited_right]
	authority: Block[0] {
		symbols: [admin, namespace, ns_operation, backlog_write, deduplication_read, retention_read, clear_backlog, offload_write, max_consumers_write, replication_write, replication_rate_read, unsubscribe, encryption_write, rate_write, ttl_read, get_permission, backlog_read, delayed_delivery_read, subscription_auth_mode_read, all_write, create_topic, delete_topic, max_unacked_write, add_bundle, anty_affinity_write, persistence_write, replication_rate_write, max_consumers_read, max_producers_read, schema_compatibility_strategy_read, deduplication_write, persistence_read, schema_compatibility_strategy_write, ttl_write, retention_write, anty_affinity_read, all_read, max_producers_write, get_topic, replication_read, encryption_read, get_bundle, get_topics, grant_permission, compaction_write, rate_read, delete_bundle, revoke_permission, compaction_read, offload_read, max_unacked_read, delayed_delivery_write, subscription_auth_mode_write, topic, topic_operation, lookup, consume, produce]
		context: 
		facts: [
			!revocation_id(Thu May 28 02:37:57 CEST 2020)
			!right(#authority, #admin)
		]
		rules: [
			*right(#authority, #namespace, $0, $1, $2) <- !ns_operation(#authority, #namespace, $0, $1, $2) @ $2 in [backlog_write, deduplication_read, retention_read, clear_backlog, offload_write, max_consumers_write, replication_write, replication_rate_read, unsubscribe, encryption_write, rate_write, ttl_read, get_permission, backlog_read, delayed_delivery_read, subscription_auth_mode_read, all_write, create_topic, delete_topic, max_unacked_write, add_bundle, anty_affinity_write, persistence_write, replication_rate_write, max_consumers_read, max_producers_read, schema_compatibility_strategy_read, deduplication_write, persistence_read, schema_compatibility_strategy_write, ttl_write, retention_write, anty_affinity_read, all_read, max_producers_write, get_topic, replication_read, encryption_read, get_bundle, get_topics, grant_permission, compaction_write, rate_read, delete_bundle, revoke_permission, compaction_read, offload_read, max_unacked_read, delayed_delivery_write, subscription_auth_mode_write]
			*right(#authority, #topic, $0, $1, $2, $3) <- !topic_operation(#authority, #topic, $0, $1, $2, $3) @ $3 in [produce, lookup, consume]
		]
		caveats: [
		]
	}
	blocks: [
		Block[1] {
		symbols: [limited_right]
		context: 
		facts: [
		]
		rules: [
		]
		caveats: [
			*limited_right(#namespace, "tenantTest", "namespaceTest", $2) <- !ns_operation(#namespace, "tenantTest", "namespaceTest", $2) @ $2 in [backlog_write, deduplication_read, retention_read, clear_backlog, offload_write, max_consumers_write, replication_rate_read, unsubscribe, encryption_write, rate_write, ttl_read, backlog_read, delayed_delivery_read, create_topic, delete_topic, max_unacked_write, add_bundle, persistence_write, max_consumers_read, max_producers_read, schema_compatibility_strategy_read, deduplication_write, persistence_read, schema_compatibility_strategy_write, ttl_write, retention_write, anty_affinity_read, all_read, max_producers_write, get_topic, replication_read, encryption_read, get_bundle, get_topics, compaction_write, rate_read, delete_bundle, compaction_read, offload_read, max_unacked_read, delayed_delivery_write] || *limited_right(#topic, "tenantTest", "namespaceTest", $2, $3) <- !topic_operation(#topic, "tenantTest", "namespaceTest", $2, $3) @ $3 in [produce, lookup, consume]
		]
	}
	]
}
[FailedCaveat.FailedBlock { block_id: 1, caveat_id: 0, rule: *limited_right(#namespace, "tenantTest", "namespaceTest", $2) <- !ns_operation(#namespace, "tenantTest", "namespaceTest", $2) @ $2 in [backlog_write, deduplication_read, retention_read, clear_backlog, offload_write, max_consumers_write, replication_rate_read, unsubscribe, encryption_write, rate_write, ttl_read, backlog_read, delayed_delivery_read, create_topic, delete_topic, max_unacked_write, add_bundle, persistence_write, max_consumers_read, max_producers_read, schema_compatibility_strategy_read, deduplication_write, persistence_read, schema_compatibility_strategy_write, ttl_write, retention_write, anty_affinity_read, all_read, max_producers_write, get_topic, replication_read, encryption_read, get_bundle, get_topics, compaction_write, rate_read, delete_bundle, compaction_read, offload_read, max_unacked_read, delayed_delivery_write] || *limited_right(#topic, "tenantTest", "namespaceTest", $2, $3) <- !topic_operation(#topic, "tenantTest", "namespaceTest", $2, $3) @ $3 in [produce, lookup, consume] }]
02:37:57.740 [main] ERROR com.clevercloud.biscuitpulsar.AuthorizationProviderBiscuit - produce verifier failure: Error.FailedLogic{ error: LogicError.FailedCaveats{ errors: [FailedCaveat.FailedBlock { block_id: 1, caveat_id: 0, rule: *limited_right(#namespace, "tenantTest", "namespaceTest", $2) <- !ns_operation(#namespace, "tenantTest", "namespaceTest", $2) @ $2 in [backlog_write, deduplication_read, retention_read, clear_backlog, offload_write, max_consumers_write, replication_rate_read, unsubscribe, encryption_write, rate_write, ttl_read, backlog_read, delayed_delivery_read, create_topic, delete_topic, max_unacked_write, add_bundle, persistence_write, max_consumers_read, max_producers_read, schema_compatibility_strategy_read, deduplication_write, persistence_read, schema_compatibility_strategy_write, ttl_write, retention_write, anty_affinity_read, all_read, max_producers_write, get_topic, replication_read, encryption_read, get_bundle, get_topics, compaction_write, rate_read, delete_bundle, compaction_read, offload_read, max_unacked_read, delayed_delivery_write] || *limited_right(#topic, "tenantTest", "namespaceTest", $2, $3) <- !topic_operation(#topic, "tenantTest", "namespaceTest", $2, $3) @ $3 in [produce, lookup, consume] }] } }
02:37:57.742 [main] DEBUG com.clevercloud.biscuitpulsar.AuthorizationProviderBiscuit - World {
	facts: [!revocation_id(Thu May 28 02:37:57 CEST 2020),
		!right(#authority, #admin)
	],
	rules: [*right(#authority, #namespace, $0, $1, $2) <- !ns_operation(#authority, #namespace, $0, $1, $2) @ $2 in [backlog_write, deduplication_read, retention_read, clear_backlog, offload_write, max_consumers_write, replication_write, replication_rate_read, unsubscribe, encryption_write, rate_write, ttl_read, get_permission, backlog_read, delayed_delivery_read, subscription_auth_mode_read, all_write, create_topic, delete_topic, max_unacked_write, add_bundle, anty_affinity_write, persistence_write, replication_rate_write, max_consumers_read, max_producers_read, schema_compatibility_strategy_read, deduplication_write, persistence_read, schema_compatibility_strategy_write, ttl_write, retention_write, anty_affinity_read, all_read, max_producers_write, get_topic, replication_read, encryption_read, get_bundle, get_topics, grant_permission, compaction_write, rate_read, delete_bundle, revoke_permission, compaction_read, offload_read, max_unacked_read, delayed_delivery_write, subscription_auth_mode_write],
		*right(#authority, #topic, $0, $1, $2, $3) <- !topic_operation(#authority, #topic, $0, $1, $2, $3) @ $3 in [produce, lookup, consume],
		*right(#authority, #topic, $0, $1, $2, #lookup) <- !right(#authority, #topic, $0, $1, $2, #produce),
		*right(#authority, #topic, $0, $1, $2, #lookup) <- !right(#authority, #topic, $0, $1, $2, #consume),
		*right(#authority, #topic, $0, $1, $2, #lookup) <- !right(#authority, #topic, $0, $1, $2, #consume, $3),
		*right(#authority, #topic, $0, $1, $2, #produce) <- !right(#authority, #namespace, $0, $1, #produce), !topic(#ambient, $0, $1, $2),
		*right(#authority, #topic, $0, $1, $2, #consume) <- !right(#authority, #namespace, $0, $1, #consume), !topic(#ambient, $0, $1, $2),
		*right(#authority, #topic, $0, $1, $2, #consume, $3) <- !right(#authority, #namespace, $0, $1, #consume), !topic(#ambient, $0, $1, $2), !subscription(#ambient, $0, $1, $2, $3),
		*right(#authority, #topic, $0, $1, $2, #produce) <- !right(#authority, #admin), !topic(#ambient, $0, $1, $2),
		*right(#authority, #topic, $0, $1, $2, #consume) <- !right(#authority, #admin), !topic(#ambient, $0, $1, $2),
		*right(#authority, #topic, $0, $1, $2, #consume, $1) <- !right(#authority, #admin), !topic(#ambient, $0, $1, $2), !subscription(#ambient, $0, $1, $2, $3)
	]
}
[FailedCaveat.FailedBlock { block_id: 1, caveat_id: 0, rule: *limited_right(#namespace, "tenantTest", "namespaceTest", $2) <- !ns_operation(#namespace, "tenantTest", "namespaceTest", $2) @ $2 in [backlog_write, deduplication_read, retention_read, clear_backlog, offload_write, max_consumers_write, replication_rate_read, unsubscribe, encryption_write, rate_write, ttl_read, backlog_read, delayed_delivery_read, create_topic, delete_topic, max_unacked_write, add_bundle, persistence_write, max_consumers_read, max_producers_read, schema_compatibility_strategy_read, deduplication_write, persistence_read, schema_compatibility_strategy_write, ttl_write, retention_write, anty_affinity_read, all_read, max_producers_write, get_topic, replication_read, encryption_read, get_bundle, get_topics, compaction_write, rate_read, delete_bundle, compaction_read, offload_read, max_unacked_read, delayed_delivery_write] || *limited_right(#topic, "tenantTest", "namespaceTest", $2, $3) <- !topic_operation(#topic, "tenantTest", "namespaceTest", $2, $3) @ $3 in [produce, lookup, consume] }]
```

I think I'm missing something to put in token or verifier as facts or rules, nope? (here starts the biscuit definition failing: https://github.com/CleverCloud/biscuit-pulsar/blob/master/biscuit-pulsar/src/test/java/com/clevercloud/biscuitpulsar/AuthorizationProviderBiscuitTest.java#L153).

**Comment by Geal** [comment]
the caveat expect a `ns_operation` fact but the only facts are `!revocation_id(Thu May 28 02:37:57 CEST 2020)` and `!right(#authority, #admin)`, that's probably the issue

**Comment by KannarFr** [comment]
Hope I fixed all we discussed in this issue on: a5c0684c46e8aa67b713d429e0fdc424a6d736fb.

WDYT @Geal?

**Comment by KannarFr** [comment]
Fixed in fde977651c6cc3054dcac29db69013d1c593f4ed.
